align_system/algorithms/decision_flow_adm/utils.py

```diff
@@ -0,0 +1,183 @@
+        max_retries: Maximum number of retry attempts (default: 3)
+        context_info: Additional context for logging (e.g., attribute name)
+
+    Returns:
+        Dict containing the parsed JSON response
+
+    Raises:
+        RuntimeError: If all retry attempts fail to generate valid JSON
+    """
+    response = None
+    last_error = None
+
+    context_str = f" for {context_info}" if context_info else ""
+
+    for attempt in range(max_retries):
+        try:
+            response = inference_engine.run_inference(
+                dialog_prompt,
+                output_schema
+            )
+
+            # Validate that response is a valid dict (JSON-like object)
+            if not isinstance(response, dict):
+                raise json.JSONDecodeError(
+                    f"Response is not a dict: {type(response)}", "", 0
+                )
+
+            # Successfully got valid JSON response
+            log.info(f"{stage_name} inference succeeded on attempt {attempt + 1}{context_str}")
+            return response
+
+        except (json.JSONDecodeError, ValueError, TypeError) as e:
+            last_error = e
+            log.warning(
+                f"{stage_name} JSON decode error on attempt {attempt + 1}/{max_retries}{context_str}: {e}"
+            )
+
+            if attempt < max_retries - 1:
+                log.info(f"Retrying {stage_name} inference{context_str}...")
+            else:
+                log.error(f"{stage_name} failed after {max_retries} attempts{context_str}")
+                raise RuntimeError(
+                    f"Failed to generate valid JSON after {max_retries} attempts{context_str}. "
+                    f"Last error: {last_error}"
+                ) from last_error
+
+    # This should never be reached due to the raise in the loop, but for type checking
+    raise RuntimeError(f"{stage_name} failed unexpectedly")
+
+
+def extract_json_from_text(text: str) -> Optional[Dict[str, Any]]:
+    """
+    Extract JSON from text that may contain markdown code blocks or other formatting.
+
+    Args:
+        text: Raw text response that may contain JSON
+
+    Returns:
+        Parsed JSON dict if found, None otherwise
+    """
+    # Try to find JSON in markdown code blocks first
+    json_patterns = [
+        r'```json\s*(\{.*?\})\s*```',  # ```json { ... } ```
+        r'```\s*(\{.*?\})\s*```',       # ``` { ... } ```
+        r'(\{.*?\})',                    # Direct JSON object
+    ]
+
+    for pattern in json_patterns:
+        match = re.search(pattern, text, re.DOTALL)
+        if match:
+            try:
+                json_str = match.group(1)
+                return json.loads(json_str)
+            except json.JSONDecodeError:
+                continue
+
+    return None
+
+
+def run_unstructured_inference_with_retry(
+    inference_engine,
+    dialog_prompt: str,
+    expected_keys: list,
+    stage_name: str,
+    max_retries: int = 3,
+    context_info: str = ""
```

Comment:
Again, not a huge fan of passing in information just for logging purposes (see previous comment)